src/Microsoft.AspNetCore.Blazor.Server/Circuits/CircuitHost.cs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Blazor.Browser.Rendering;
+using Microsoft.AspNetCore.Blazor.Services;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Blazor.Server.Circuits
+{
+    internal class CircuitHost : IDisposable
+    {
+        public event UnhandledExceptionEventHandler UnhandledException;
+
+        private bool _isInitialized;
+        private Action<BrowserRenderer> _configureRenderer;
+
+        public CircuitHost(
+            IServiceScope scope,
+            BrowserRenderer renderer,
+            Action<BrowserRenderer> configureRenderer,
+            IJSRuntime jsRuntime,
+            CircuitSynchronizationContext synchronizationContext)
+        {
+            Scope = scope ?? throw new ArgumentNullException(nameof(scope));
+            Renderer = renderer ?? throw new ArgumentNullException(nameof(renderer));
+            _configureRenderer = configureRenderer ?? throw new ArgumentNullException(nameof(configureRenderer));
+            JSRuntime = jsRuntime ?? throw new ArgumentNullException(nameof(jsRuntime));
+            SynchronizationContext = synchronizationContext ?? throw new ArgumentNullException(nameof(synchronizationContext));
+
+            Services = scope.ServiceProvider;
+
+            Circuit = new Circuit(this);
+        }
+
+        public Circuit Circuit { get; }
+
+        public IJSRuntime JSRuntime { get; }
+
+        public BrowserRenderer Renderer { get; }
+
+        public IServiceScope Scope { get; }
+
+        public IServiceProvider Services { get; }
+        
+        public CircuitSynchronizationContext SynchronizationContext { get; }
+
+        public async Task InitializeAsync()
+        {
+            await SynchronizationContext.Invoke(() =>
+            {
+                Microsoft.JSInterop.JSRuntime.SetCurrentJSRuntime(JSRuntime);
+                _configureRenderer(Renderer);
+            });
+
+            _isInitialized = true;
+        }
+
+        public async void BeginInvokeDotNetFromJS(string callId, string assemblyName, string methodIdentifier, string argsJson)
+        {
+            AssertInitialized();
+
+            try
+            {
+                await SynchronizationContext.Invoke(() =>
+                {
+                    Microsoft.JSInterop.JSRuntime.SetCurrentJSRuntime(JSRuntime);
+
+                    switch (methodIdentifier)
+                    {
+                        // Massive hack. Need a common system for dispatching calls within the context of
```

Comment:
I just copied your hack for now. We'll need to solve this or add interop for instances.

Reply:
When we receive incoming JS interop calls to .NET, is there any way to access the scoped service provider? I expect there will need to be, e.g., for having an `AppState` scoped to a circuit.

If it is possible to do that, then `NotifyLocationChanged` can invoke some static method that accesses the scoped `RemoteUriHelper` to call its instance method.

Reply:
We could try something like that. I was planning to make it possible to access the Circuit via async-local, which accomplises the same thing - basically what you suggested in your comment here.

Reply:
The tricky part btw is what to do in the client - does it invoke a different static method for serversideblazor?

Reply:
What if we enhance the JS interop mechanism so that, if you're trying to do a JSON-marshalled call to a .NET method that takes a parameter of type `IServiceProvider`, then for that parameter we automatically supply the one that corresponds to:

 * Your circuit, for server-side execution
 * The whole Blazor application, for WebAssembly execution

I don't think this would break any other scenario, since it's not possible to supply an `IServiceProvider` instance through JSON-marshalled calls in any other way. Once the .NET method receives the `IServiceProvider`, it can acquire the circuit or IJSRuntime or whatever else from it.

Or if we wanted to make it nicer still, we could look at supporting a `[FromServices]` attribute that you can put on `[JSInvokable]` methods' parameters, and then we supply whatever type of thing from your current service provider.

Reply:
I'm not totally sure that makes sense as a feature of interop writ-large. It seems awfully tied to our purposes. The other option would be to just have the JS-side call different .NET methods depending on the environment. 

But we do have the general problem right now where everything we need to to do via interop needs to go through statics. 

We can either address it on the JS side by calling different methods, or on the .NET side by building more magic-statics. If we have the JS side call different methods we still need magic statics to some degree.

This gets much better if we implement interop for instances.

Reply:
I'm going to start by enabling navigation interception earlier in the handshake and see where this leads me.